objectserver/nurserystabilizer.go

```diff
@@ -0,0 +1,218 @@
+			nrd.r.nurseryConcurrencySem <- struct{}{}
+			defer func() {
+				<-nrd.r.nurseryConcurrencySem
+			}()
+			if ok, err := nrd.CanStabilize(o); ok {
+				if err = o.Stabilize(); err == nil {
+					nrd.updateStat("objectsStabilized", 1)
+					nrd.notifyPeers(o)
+				} else {
+					nrd.r.logger.Error("[stabilizeDevice] error Stabilize obj", zap.String("Object", o.Repr()), zap.Error(err))
+				}
+			} else {
+				if err != nil {
+					nrd.r.logger.Error("[stabilizeDevice] error CanStabilize obj", zap.String("Object", o.Repr()), zap.Error(err))
+				}
+			}
+		}()
+		select {
+		case <-time.After(nurseryObjectSleep):
+		case <-nrd.canchan:
+			return
+		}
+	}
+	nrd.updateStat("PassComplete", 1)
+}
+
+func (nrd *nurseryDevice) notifyPeers(o ObjectStabilizer) error {
+	metadata := o.Metadata()
+	ns := strings.SplitN(metadata["name"], "/", 4)
+	if len(ns) != 4 {
+		return fmt.Errorf("invalid metadata name: %s", metadata["name"])
+	}
+	partition := nrd.oring.GetPartition(ns[1], ns[2], ns[3])
+	if _, handoff := nrd.oring.GetJobNodes(partition, nrd.dev.Id); handoff {
+		return nil
+	}
+	nodes := nrd.oring.GetNodes(partition)
+	for _, device := range nodes {
+		if device.Ip == nrd.dev.Ip && device.Port == nrd.dev.Port && device.Device == device.Device {
+			continue
+		}
+		url := fmt.Sprintf("http://%s:%d/stabilize/%s/%d%s", device.Ip, device.ReplicationPort, device.Device, partition, common.Urlencode(metadata["name"]))
+		if req, err := http.NewRequest("POST", url, nil); err == nil {
+			req.Header.Set("X-Backend-Storage-Policy-Index", strconv.FormatInt(int64(nrd.policy), 10))
+			req.Header.Set("X-Backend-Nursery-Stabilize", "true")
+			req.Header.Set("X-Timestamp", metadata["X-Backend-Data-Timestamp"])
+			req.Header.Set("User-Agent", "nursery-peer-stabilizer")
+			resp, _ := nrd.client.Do(req)
+
+			if resp != nil {
+				resp.Body.Close()
+			}
+		} else {
+			return err
+		}
+	}
+	return nil
+}
+
+func (nrd *nurseryDevice) CanStabilize(o ObjectStabilizer) (bool, error) {
+	metadata := o.Metadata()
+	ns := strings.SplitN(metadata["name"], "/", 4)
+	if len(ns) != 4 {
+		return false, fmt.Errorf("invalid metadata name: %s", metadata["name"])
+	}
+	partition := nrd.oring.GetPartition(ns[1], ns[2], ns[3])
+	if _, handoff := nrd.oring.GetJobNodes(partition, nrd.dev.Id); handoff {
+		return false, nil
+	}
+	nodes := nrd.oring.GetNodes(partition)
+	goodNodes := uint64(0)
+	for _, device := range nodes {
+		if device.Ip == nrd.dev.Ip && device.Port == nrd.dev.Port && device.Device == device.Device {
+			continue
+		}
+		url := fmt.Sprintf("http://%s:%d/%s/%d%s", device.Ip, device.Port, device.Device, partition, common.Urlencode(metadata["name"]))
+		req, err := http.NewRequest("HEAD", url, nil)
+		req.Header.Set("X-Backend-Storage-Policy-Index", strconv.FormatInt(int64(nrd.policy), 10))
+		req.Header.Set("User-Agent", "nursery-stabilizer")
+		resp, err := nrd.client.Do(req)
+
+		if err == nil && (resp.StatusCode/100 == 2 || resp.StatusCode == 404) &&
```

Comment:
I'm not sure I understand why 404 is ok here.

Reply:
tombstones will 404

Reply:
What about "I don't have that object at all"? Isn't that important here?

Reply:
if you don't have it at all then goodNodes won't be incremented and the object will be skipped. it relies on replication to populate the other nodes.